### ISSUE TITLE
Provide an extension for MapStruct

### ISSUE THREAD

**Issue by gunnarmorling** [quarkusio/quarkus#5956]
**Describe the extension**

MapStruct is a compile-time code generator (JSR 269 annotation processor) for bean-to-bean-mappings, useful e.g. to map an internal (entity) model to the external (REST) model of an application. As the generated code avoids any kind of reflection, MapStruct leans itself to being a perfect match with Quarkus and GraalVM.

Interested in this extension, please +1 via the emoji/reaction feature of GitHub (top right).

**Configuration suggestion**

MapStruct has a few annotation processor options that impact the generated source code (e.g. whether to include or not a timestamp). I suppose they might be made static Quarkus properties, but not sure whether that's needed really. To clarify, these options are not examined by the generated code, but they impact *how* the code is generated.

**Additional context**

MapStruct works with Quarkus/GraalVM out-of-the-box. So the value of the extension would be better a getting started experience by means of selecting it from the catalogue. One thing it actually could do is to enable all generated mappers for reflection. That's not needed when using CDI as the component model for the generated code (which is what I'd recommend with Quarkus), but there's an alternative way to obtain mapper instances which uses reflection (`Mappers.getInstance(MyMapper.class)`). This is the only location where reflection is used so it'd benefit from automatic enabling this.

**Comment by gunnarmorling** [comment]
Hey @chris922, together with @gsmet and @Sanne we were discussing to have an official Quarkus extension for MapStruct. As you nicely point out in your post, MapStruct really works out of the box here, but such extension would still make for a nicer getting started experience and people could choose it from the extension catalogue etc. Would you perhaps be interested in leading this effort?

**Comment by gunnarmorling** [comment]
Btw. a related yet general question is: how should annotation processors such as MapStruct or the JPA metamodel generator integrated into Quarkus? I was thinking whether it'd make sense to have a JSR 269 `Filer` implementation which takes the generated source code and builds byte code right away. Just thinking out loud, perhaps it doesn't add value.

**Comment by Sanne** [comment]
@gunnarmorling we don't dislike annotation-processors; as you pointed out during one of the brainstorming meetings, while their API is a hard nut they provide nice integration with the IDEs, allowing auto-completion to work on the fly: seems fitting for MapStruct and JPA metamodel.

I think we should keep that - unless I misunderstood what you have in mind?

**Comment by gunnarmorling** [comment]
> I think we should keep that

Ok, cool. I somehow had the impression that you'd want to move away from annotation processors, but it's good to know that this isn't the case.

Btw. a related issue in this context is https://github.com/quarkusio/quarkus/issues/1502, which prevents from APs to be re-run during Dev mode in certain set-ups. Would be nice to have this addressed, as currently the set-up we recommend (using the `annotationProcessPaths` option of the Maven compiler plug-in) isn't working with Dev mode, so the AP JAR must be added as a dependency instead which isn't ideal.

**Comment by Sanne** [comment]
> Ok, cool. I somehow had the impression that you'd want to move away from annotation processors, but it's good to know that this isn't the case.

Following your advice :)

Regarding #1502 I don't know, need to leave that to others. Sad to see it's rather old and yet unaddressed.

**Comment by maxandersen** [comment]
annotation processors are generally slow compared to a more direct quarkus extension so if we can avoid them it would be great.

another issue I keep having with mapstruct is that it requires my fields to be public where we don't need elsewhere. could a proper mapstruct extension remedy that ?

**Comment by filiphr** [comment]
>annotation processors are generally slow compared to a more direct quarkus extension so if we can avoid them it would be great.

I am not entirely sure how Quarkus extensions works. However, shouldn't the annotation processor eventually be involved somehow? How would the generated code be invoked otherwise?

>another issue I keep having with mapstruct is that it requires my fields to be public where we don't need elsewhere. could a proper mapstruct extension remedy that ?

I would be curious to understand why you think that mapstruct needs the fields to be public. MapStruct does not have such a requirement. You need to have public getters / setters. There is https://github.com/mapstruct/mapstruct/issues/1689 which asks for having the ability to use non public visibility. If that's what you are looking for, I do not think that a Quarkus extension will help, since it is something that we explicitly check for in the MapStruct annotation processor.

**Comment by Sanne** [comment]
> annotation processors are generally slow compared to a more direct quarkus extension so if we can avoid them it would be great.

There are some situations in which we prefer an AP for code generation: the cases in which it's useful to have immediate feedback within the IDE.

I believe this is one of such cases so it might be worth sticking to the AP approach.

**Comment by maxandersen** [comment]
@filiphr I'm talking about fields, especially package private scoped not needing public accessors. Yes I seen the #1689 from 2019 thus I assumed there were some technical reason (like having to use reflection in the generated source) for why it wasn't there.

Hence I just wondered if the mapping implementations were code generated during build (rather than in compile phase) mapstruct could do more. 

@Sanne maybe i'm missing some feature of mapstruct but how is this different from i.e. hibernate validating errors at build time ? with quarkus:dev you get errors too while coding.

**Comment by filiphr** [comment]
>Yes I seen the https://github.com/mapstruct/mapstruct/issues/1689 from 2019 thus I assumed there were some technical reason (like having to use reflection in the generated source) for why it wasn't there.

It is more of a maintenance thing from our side, we need to do some refactorings to make sure we have access to package that the class will be generated to be able to determine whether we can do the mapping or not. It has nothing to do with reflection since MapStruct uses no reflection to do the mapping.

**Comment by geoand** [comment]
> I'm currently annoyed because I'm getting comfortable using quarkus:dev mode and mapstruct processor does'nt "regenerate" on the fly the modification made to the mapper.

Is this still the case?

**Comment by fordsworth** [comment]
> > I'm currently annoyed because I'm getting comfortable using quarkus:dev mode and mapstruct processor does'nt "regenerate" on the fly the modification made to the mapper.
> 
> Is this still the case?

Yes, still an issue on 3.29.0. You literally have run mvn compile quarkus:dev

**Comment by Postremus** [comment]
Same issue here.

@fordsworth You can also force recreation of the mapper. Just add a new line in the mapper interface / abstract class. That class will then be rebuild, causing the mapstruct processor to create a new mapper implemention.

But, that is obviously way to cumbersome to remember, It is easier to restart the whole project.. Which defeats the purpose of dev mode.

The problem here is, in general, that a class gets changed, but quarkus is not able to determine that another related class should also be marked for recompilation.

A simple (ish) extension for mapstruct which solves this problem could (imo) look like this:
* Search for all interfaces / classes annotated with `Mapper`
* Scan all theirs methods
* Produce a a build item. This build item could let quarkus know "hey, if class x will be recompiled, please recompile class y afterwards as well".

The first two steps are already possible in an extension, I however do not see a way to for the last step. (On a first glance) there are no hooks in RuntimeUpdatesProcessor to expand which classes should be recompiled.

@geoand Do you agree with this general implementation plan?

**Comment by geoand** [comment]
> there are no hooks in RuntimeUpdatesProcessor to expand which classes should be recompiled.

Correct.

> @geoand Do you agree with this general implementation plan?

It does sounds reasonable

**Comment by Postremus** [comment]
Assigning myself for now. 
In the first step, I will create a prototype for https://github.com/quarkusio/quarkus/issues/5956#issuecomment-3479342534 . (Extension part and quarkus core parts.)

**Comment by gsmet** [comment]
My understanding is that we added some integration for the annotation processors. What’s missing is that we don’t detect something has changed and we don’t run the AP?

**Comment by Postremus** [comment]
@gsmet  yes.

For example, take the following models (get / set ommited for brevity):

```
class A {

String firstname;

}

class B {

String lastname;

}

```

And the mapper:
```
interface AMapper {

@Mapper
B mapToB(A a);

}
```

Now when we run `mvn quarkus:dev`, class A, B, and the AMapper are compiled. During this, the Annotation processor from mapstruct runs, and generates an Implementation for AMapper. So far so good.

Now lets assume, we expand A and B with a property `lastname`.
We trigger a hot reload using rest.
Only A and B get recompiled, however the Annotation Processor does not get triggered, since we do not recompile AMapper.

Therefore, the mapper does not map the newly added `lastname` property until either AMapper changes, or quarkus dev mode is restarted.

**Comment by filiphr** [comment]
@Postremus, not sure what kind of integration you have for annotation processors in general. What we at MapStruct do is that we use the [`Filler#createSourceFile`](https://docs.oracle.com/en/java/javase/17/docs/api/java.compiler/javax/annotation/processing/Filer.html#createSourceFile(java.lang.CharSequence,javax.lang.model.element.Element...)) and in the `originatingElements` we pass the mapper class. e.g. `AMapper` in this the previous example.

Ideally quarkus should detect the dependencies of this class and if something in those dependencies changes it should trigger the annotation processing for that mapper again. This is not linked to MapStruct at all, it's a general thing that every annotation processor does.

I personally have not used quarkus, so I do not know how it works. I know that IntelliJ and Gradle intercept that `Filer` and create their own dependency graph for classes and know when to trigger it. I know this since I've interacted with them for certain improvements in their incremental processing support. 

Please let me know, if there is something from MapStruct side you need? I'm the project lead of the project in the moment.

**Comment by Postremus** [comment]
I got the basic case (from https://github.com/quarkusio/quarkus/issues/5956#issuecomment-3481641653) working.

If I have changes in class A, and trigger a hot reload, then AMapper is also recompiled, and a new AMapperImpl is generated.

Right now I do this by directly integrating with mapstruct by implementing `org.mapstruct.ap.internal.processor.ModelElementProcessor` in a separate quarkus-mapstruct-processor artifact.

In the next few steps, I want to 
1) clean up my code - it is looking really rough.
2) Attempt a more generic way. I (believe) that we won't be able to integrate deeply with generic annotation processors using [Filler#createSourceFile](https://docs.oracle.com/en/java/javase/17/docs/api/java.compiler/javax/annotation/processing/Filer.html#createSourceFile(java.lang.CharSequence,javax.lang.model.element.Element...)) as mentioned by @filiphr. Mostly because I see no existing infrastructure on our side to accomplish this. (And I'm not tot knowledgable with annotation processors myself)
However, what we could do is a search using jandex for all the method parameter types in the mapper interface, and translate those types back to source file locations.
We might be able to leverage this more generic way later on for other annotation processors, which (could) have a similar problem.
Immutables comes to mind.


@filiphr 
Do you by any chance have example on hand which uses all the mapstruct options?


@geoand 
By any chance, do you know about a premade utility in the quarkus-core to translate fqdn (com.acme.A) to a source file location (C:\folder\project\src\main\java\com\acme\A.java)? Right now I use the `CurateOutcomeBuildItem`. Works but is a bit clunky.

**Comment by geoand** [comment]
@Postremus nice!

The only thing I remember is [this](https://github.com/quarkusio/quarkus/blob/2806298182ff11a058073412991193a8c93dc801/core/deployment/src/main/java/io/quarkus/deployment/ide/Ide.java#L98-L114)

**Comment by Postremus** [comment]
I tested using jandex for mapper discovery, and recompilation triggers. Works even better then directly integrating with mapstruct.

Implementing `org.mapstruct.ap.internal.processor.ModelElementProcessor` had 1 big advantage: 
* I could reuse mapstructs  `GeneratedTypeMethod`s and their type parameters. These methods include every methods (afaik) that is supposed to be generated in the final mapper impl.

However, some drawbacks:
* I had to run this implementation alongside the mapstruct-processor. How do I get the discovered recompilation triggers back into the build step? Only solution I could come up with for now was writing down a small discovery-file.properties.
* This implementation was only run, when the mapstruct processor was also run. I.e. to refresh the discovery-file.properties, a user needed to recompile the mapping class. So the file itself was potentially outdated. 
* And the last, and biggest problem, during the first quarkus:dev run, mapstruct returned a simple class Name, instead of the FQDN like I asked it to. When triggering recompilation through changes in the mapper class, the FQDN was returned again.
Is this something I should open a bug for in mapstruct @filiphr ? I.e. instead of `com.acme.Address`, I got simply `Address` with this code.
```
public Mapper process(ProcessorContext context, TypeElement typeElement, Mapper mapper) {

        Set<String> importTypeNames = new HashSet<>();
        for (GeneratedTypeMethod method : mapper.getMethods()) {
            for (Type importType : method.getImportTypes()) {
                importTypeNames.add(importType.getFullyQualifiedName());
            }
        }
```


@geoand 
I would therefore go ahead with the jandex based approach, which seems to work fine enough.

My proposal based on this initial prototype is to:
* create a quarkiverse-mapstruct extension
** This extension determines the Mapper classes, and their recompilation triggers.
** Produces Build Items, which tells quarkus core: if class FQDN changes, also recompile this FQDN
* quarkus-core 
** Provides this Build
** Consume this BuildItem and translates these FQDN to source locations. 
** RuntimeUpdatesProcessor gets expanded to actually do the recompilation of the mapper class

This way, If the need arises to do something similar for another annotation processor, we would "just" have to produce the according builditems.


Do you agree with this approach, and could you please create the quarkiverse extension?

**Comment by geoand** [comment]
Thanks a lot @Postremus.

I would like to have @FroMage 's opinion, as he has had to deal with annotation processors as well in the past and might have some additional insights.

**Comment by filiphr** [comment]
>Do you by any chance have example on hand which uses all the mapstruct options?

The number of options is quite complex. Some more complex ones can be found in our test suite:

* https://github.com/mapstruct/mapstruct/tree/main/processor/src/test/java/org/mapstruct/ap/test/factories
* https://github.com/mapstruct/mapstruct/tree/main/processor/src/test/java/org/mapstruct/ap/test/inheritedmappingmethod
* https://github.com/mapstruct/mapstruct/tree/main/processor/src/test/java/org/mapstruct/ap/test/inheritance
* https://github.com/mapstruct/mapstruct/tree/main/processor/src/test/java/org/mapstruct/ap/test/imports
* https://github.com/mapstruct/mapstruct/tree/main/processor/src/test/java/org/mapstruct/ap/test/mapperconfig
* https://github.com/mapstruct/mapstruct/tree/main/processor/src/test/java/org/mapstruct/ap/test/nestedbeans

>Implementing `org.mapstruct.ap.internal.processor.ModelElementProcessor` had 1 big advantage:

That one is a bit internal, but could work. Although if we change something then it'll break.


>And the last, and biggest problem, during the first quarkus:dev run, mapstruct returned a simple class Name, instead of the FQDN like I asked it to. When triggering recompilation through changes in the mapper class, the FQDN was returned again.
Is this something I should open a bug for in mapstruct @filiphr ? I.e. instead of com.acme.Address, I got simply Address with this code.

I'll need a reproducal to investigate why this is happening. It could be that there is some file not yet ready so we don't really generate the mappers in the first round of the annotation processing.

----

In any case, I still believe that the better approach is to do something on the lower level, since then you have a solution for all annotation processors, not just for MapStruct.

I checked out the Quarkus code base and I did find some things that might work a bit better for Quarkus. I started from the `DevMojo` and that eventually led me to [`JavaCompilationProvider`](https://github.com/quarkusio/quarkus/blob/main/core/deployment/src/main/java/io/quarkus/deployment/dev/JavaCompilationProvider.java). Here I could see that you are actually the one triggering the `JavaCompiler`. If we follow what `Filer#createSourceFile` we end up at [`JavaFileManager#getJavaFileForOutput`](https://docs.oracle.com/en/java/javase/17/docs/api/java.compiler/javax/tools/JavaFileManager.html#getJavaFileForOutput(javax.tools.JavaFileManager.Location,java.lang.String,javax.tools.JavaFileObject.Kind,javax.tools.FileObject)). Which unfortunately is a bit buggy on Java 17, since it does not pass the `sibling` properly. However, from Java 18 this has been fixed as part of https://bugs.openjdk.org/browse/JDK-8272234. 

Long story short, perhaps it is worth exploring the [`QuarkusFileManager`](https://github.com/quarkusio/quarkus/blob/main/core/deployment/src/main/java/io/quarkus/deployment/dev/filesystem/QuarkusFileManager.java). I don't know whether it is possible to hook the creation of `BuildItem`(s) on this level. If you can, then in the `RuntimeUpdatesProcessor` you will most likely need to find the classes that have some of those dependents as changed and pass them as to the compiler. Right now you are only passing classes that changed, but not classes that could lead to changes in annotation processors.

**Comment by Postremus** [comment]
> I'll need a reproducal to investigate why this is happening. It could be that there is some file not yet ready so we don't really generate the mappers in the first round of the annotation processing.

Problem seems to be gone after recompiling my project. Sorry about the noise.


> I don't know whether it is possible to hook the creation of BuildItem(s) on this level. If you can, then in the RuntimeUpdatesProcessor you will most likely need to find the classes that have some of those dependents as changed and pass them as to the compiler.

Yes, that is exactly what I am doing with the jandex Index. I figure out all the `@Mapper` annotated classes, and note down the dependents (in my previous comment I called them recompilation triggers..). The mapping depentens -> mapper classes is then passed down to the `RuntimeUpdatesProcessor `.

> In any case, I still believe that the better approach is to do something on the lower level, since then you have a solution for all annotation processors, not just for MapStruct.

What we could do, is to make the annotation(s) we search for configurable. 
In this case, my proposal changes slightly:
* quarkus-core
** Provide a config, where all the annotations we should search for can be specified in.
**  Determine the annotated classes, and their dependents.
** Produces Build Items, which tells quarkus core: if class FQDN changes, also recompile this FQDN
** Consume this BuildItem and translates these FQDN to source locations.
** RuntimeUpdatesProcessor gets expanded to actually do the recompilation of the mapper class
* create a quarkiverse-mapstruct extension
** Provide a default config for the annotations we search for, org.mapstruct.Mapper

> I checked out the Quarkus code base and I did find some things that might work a bit better for Quarkus. I started from the DevMojo and that eventually led me to [JavaCompilationProvider](https://github.com/quarkusio/quarkus/blob/main/core/deployment/src/main/java/io/quarkus/deployment/dev/JavaCompilationProvider.java). Here I could see that you are actually the one triggering the JavaCompiler.

Sometimes I guess yes, we are triggering the compilation ourselves. When the class file has not been compiled yet, or if there are changes to the source file for hot reload.
In modules without the quarkus plugin, the maven-compiler plugin compiles the sources for us. 
Also, if e.g. `mvn clean compile quarkus:dev` is invoked instead of just the dev goal, then the maven compiler does the compilation as well.
In both of these cases the `QuarkusFileManager` would not be invoked / be able to retrieve the originatingElements.

**Comment by filiphr** [comment]
>Yes, that is exactly what I am doing with the jandex Index.

In theory to provide a more native support with MapStruct, we can look into having a `ModelElementProcessor` that is going to output this jandex index. We'll need some more info for how this should look like and we can output it in the same process we are outputting the generated code. By the way, you can't use `getImportTypes()` because this one only returns the types that should be imported on the generated class. However, if you have classes that have the same simple name, but different FQN, then only one of those would be in the `getImportTypes()`.

>What we could do, is to make the annotation(s) we search for configurable.
>In this case, my proposal changes slightly:
>
>quarkus-core
>** Provide a config, where all the annotations we should search for can be specified in.

I would actually do this differently. There should be no need to provide a config. You can detect this automatically. You could perhaps use [`Processor#getSupportedAnnotationTypes`](https://docs.oracle.com/en/java/javase/17/docs/api/java.compiler/javax/annotation/processing/Processor.html#getSupportedAnnotationTypes()) to automatically detect which annotations you should detect.

>Sometimes I guess yes, we are triggering the compilation ourselves. When the class file has not been compiled yet, or if there are changes to the source file for hot reload.
In modules without the quarkus plugin, the maven-compiler plugin compiles the sources for us.

Thanks for explaining this to me, I wasn't aware about it.

**Comment by FroMage** [comment]
I agree that using the `Filer` to obtain the dependencies of the annotation processor targets would be a safer fit, but as @Postremus noted, we're not in control of the first compilation done by Maven, so we would miss those at startup, and only obtain them when the annotation processor ran, requiring an initial change in the annotated class.

So if there's an easy way to use Jandex to build those dependencies ourselves, then that's a good solution, and it sounds like it's pretty straightforward, so @Postremus 's strategy sound good to me 👍

**Comment by Postremus** [comment]
The first part is available as PR #51145. This resolves the problem, that the mapper is not automatically regenerated on hot reload when a model class changes. This requires a manuel configuration for now though.
As described here: https://github.com/quarkusio/quarkus/issues/5956#issuecomment-3481641653

Next part required for this ticket, as described by in the op:

> MapStruct works with Quarkus/GraalVM out-of-the-box. So the value of the extension would be better a getting started  experience by means of selecting it from the catalogue. One thing it actually could do is to enable all generated mappers for reflection. That's not needed when using CDI as the component model for the generated code (which is what I'd recommend with Quarkus), but there's an alternative way to obtain mapper instances which uses reflection (Mappers.getInstance(MyMapper.class)). This is the only location where reflection is used so it'd benefit from automatic enabling this.

I have a small prototype for this already finished. It was enough to register the mappers for reflection. Might need to register a few more supporting classes, tests will show this.

@geoand
Do you see value in having this as a separate quarkiverse mapstruct extension; and if so, could you prepare the repo?
We could also auto enable #51145 for mapstruct in this extension, so the user wouldn't have to configure anything.

**Comment by geoand** [comment]
@Postremus that makes sense to me! @gastaldi handles the quarkiverse repos
